### DISCUSSION TITLE
macOS Desktop App with Mac Catalyst and/or AppKit

### DISCUSSION THREAD

**Discussion by EtnaTraining** [appcelerator/titanium_mobile#12607] [Q&A]
Hello,

from the release note of the Titanium SDK 9.2.0 [(link)](https://wiki.appcelerator.org/display/guides2/Titanium+SDK+9.2.0+RC+Release+Note#TitaniumSDK9.2.0RCReleaseNote-macOSsupport) I read about an initial support for building desktop app on macOS using Mac Catalyst. 

What is the status of this effort? 
I am also wondering if using Hyperloop we could access AppKit native APIs

It would be really cool to have macOS desktop support in Titanium. 
As far as I know, no other cross platform desktop environment offers native API support for desktop from JS.
Just for a reference of platforms that promise desktop support:
- Flutter 2, very new, but it use to draw custom widget (no platform UI elements)
- react-native-mac, supported by Microsoft, it exposes some basic APIs, no way to access directly AppKit APIs from JS
- NativeScript has the technology (similar to Hyperloop) to access native APIs, they have proposal on that, but no decision taken
- Xamarin.Mac, it provides access to AppKit APIs, but from C#
- NodeGUI (https://github.com/nodegui/nodegui), but it exposes QT5 widgets (that are not native)

So Titanium MacOS Catalyst it seems to me the first and only framework that provides a way to build macOS native apps.

Is there any relevant code to look at in the repository to see what's the current status?

**Comment by jquick-axway** [comment]
Like you said, Titanium's mac app support is implemented via MacOS Catalyst. You can build a mac app via...
```
appc run -p ios -T macos
```
...or...
```
ti build -p ios -T macos
```

We do actively run our unit test suite on Titanium built mac apps every time a PR is submitted.

Admittedly, we don't actively develop for it other than to keep it alive and to ensure our iOS features/changes continue to work on mac. The last Mac specific feature we added was checkbox support to the upcoming 10.0.0 (see: [TIMOB-28302](https://jira.appcelerator.org/browse/TIMOB-28302)).

**Reply by EtnaTraining** [reply]
thank you for the reply.
Does macOS target support the usage of modules extension mechanism ?
Is there any source code to study to get it started in that case?

**Reply by jquick-axway** [reply]
> Does macOS target support the usage of modules extension mechanism ?

Yes, but only modules that are built with Titanium 9.2.0 or higher. That and the library dependencies a module references have to support Mac catalyst as well. I'm able to build and run sample code [kitchensink-v2](https://github.com/appcelerator/kitchensink-v2) as a mac app just fine... except... I have to exclude its "ti.facebook" module since it currently has build issue with Mac catalyst. The "ti.map" module works on Mac though.

My only word of warning is that Mac app support is not "actively" supported. It's "passively" supported via unit testing. It's considered "beta" and it will likely stay that way unless the community supports it. On our end, Android and iOS is top priority.

**Reply by EtnaTraining** [reply]
that sounds great!
ok, I understand that macOS support is not your top priority, I am just exploring the current capabilities. Thank you.

Last question, is also Hyperloop supposed to work with the macOS target? In that case, maybe any missing/not supported features could be implemented with it (I am thinking about the menubar configuration for example...), apart from modules (modules can also developed via Swift now?)

**Reply by EtnaTraining** [reply]
never mind the last question: I tried and it seem Hyperloop works also on macOS target!!! This is great!
However it seems not working with Liveview enabled (but I am not sure if Liveview works for regular iOS and Android app that makes use of Hyperloop)

So having both Hyperloop and Module support, any current limitation of the macOS target support could be implemented with one of those technologies or are there any other limitation? For example I can add the menu and menubar configuration?)

**Reply by jquick-axway** [reply]
iOS has LiveView/Hyperloop issues as well. It's not an issue on Android as of Titanium 9.0.0 because we "accidentally" fixed it when we refactored everything for gradle. It was one of those Bob Ross happy little accidents. 😛 

So as far as limitations, there are some Mac exclusive features that you're missing. Things like setting up the app's menu, no right click events, mouse hover events, etc. that might be good to have. I don't think we have M1 architecture support for it either which might be the bigger concern.

**Reply by acaland** [reply]
but if Hyperloop is supported on the macOS Catalyst target, would it be possibile to implement any missing features with it? In principle via Hyperloop I could access from JS the Menu app APIs. Would it be possible or other issues I don's see could arise?

**Reply by jquick-axway** [reply]
Yes, it's possible to implement missing features via hyperloop... and via native modules as well.
